src/System.Linq.Queryable/src/System/Linq/Queryable.cs

```diff
@@ -842,6 +842,23 @@ public static TSource SingleOrDefault<TSource>(this IQueryable<TSource> source,
                     ));
         }
 
+        public static bool TrySingle<TSource>(this IQueryable<TSource> source, out TSource element)
+        {
+            Tuple<bool, TSource> result = TrySingle(source);
+            element = result.Item2;
+            return result.Item1;
+        }
+
+        internal static Tuple<bool, TSource> TrySingle<TSource>(this IQueryable<TSource> source)
+        {
+            if (source == null)
+                throw Error.ArgumentNull(nameof(source));
+            return source.Provider.Execute<Tuple<bool, TSource>>(
+                Expression.Call(
+                    null,
+                    CachedReflectionInfo.TrySingle_TSource_1(typeof(TSource)), source.Expression));
+        }
+
```

Comment:
One more note on my observation above.

The fact that `IQueryable<T>` implements `IEnumerable<T>` makes it equivalent to a `Func<IEnumerator<T>>` by means of having the `GetEnumerator` method. So it really *is* a `Compile` method in disguise, concretely implemented by the query provider author behind the `GetEnumerator` (and sometimes more deferred, behind the first `MoveNext` method, through an iterator) by passing the `Expression` to the `Execute` method.

In practice, the execution often involves the creation of a `Func<IEnumerable<T>>` through expression compilation etc., to include a result materialization step (e.g. the "final" `Select` clause causing object creation) and/or a partial local evaluation. So it really is using `Execute` as if there's a `Compile` step returning a compiled execution plan. This is often accompanied by other tricks to be able to reuse the same "plan" (by lifting out parameters that may change, e.g. from locally evaluated closures, and caching the resulting `Func<IEnumerable<T>>`) across multiple enumerations over the same query result.

This just shows that the aggregate methods really short-circuit a lot of phasing that's otherwise available through the laziness of `IEnumerable<T>`. For example:

```csharp
IQueryable<int> res = xs.Where(...).Select(...);
while (true)
{
int sum = res.Sum();
}
```
Short of playing caching tricks deep inside the provider, repeated evaluation of `Sum` does not give the query provider an opportunity to cache the "execution plan". It will be given a new expression tree each and every time. Contrast this to:

```csharp
IQueryable<int> res = xs.Where(...).Select(...);
while (true)
{
    foreach (int x in res) { ... }
}
```
In here, every repeated enumeration of result sequence can leverage the same plan which can be stored with the `IQueryable<T>` instance in `res`. This is a trick I play quite often in the stuff I'm working on here in Bing. The first `GetEnumerator` can prepare/compile the plan; subsequent ones can leverage it again. It does so by lifting out pieces that have to be re-evaluated locally each time (such as closed-over variables) through "funclet" expressions and generating a parameterized `Func<..., IEnumerator<T>>` factory which encapsulates the cached template that will be sent to the service over and over again (along with freshly evaluated arguments from the "funclets") upon repeated evaluation.

Alas, `Execute` does not provide an such option, because it doesn't return a `Quote<int>` like type (which would have an `Evaluate` method to get the `int`, and thus would be an object that can cache the "prepared" execution plan). Short of generalized quotes, even an `Execute` that'd return a `Func<T>` (or more generally, as described above, a `TDelegate`) would result in an object that can encapsulate cached execution plans for repeated evaluation (though it'd lack the ability to do further composition of the result of aggregates):

```csharp
Func<int> sum = res.Where(...).Select(...).Sum();
while (true)
{
    sum(); // no re-processing of a newly created expression tree over and over again
}
```

Also adding @divega who can chime in on "tales from the Entity Framework trenches" and offer an opinion on which way we should take `Queryable.TrySingle` so they can implement support for it.